### DISCUSSION TITLE
Allow actions/activities to be defined for *any* state

### DISCUSSION THREAD

**Discussion by simonihmig** [statelyai/xstate#3463] [Q&A] [answered]
I am trying to expose a machine as part of a library. The state logic it defines should be applicable to any consumer. But the side effects can vary wildly, so basically the user should be able to define actions (and activities/invokable callbacks) for any (entry or exit) of any state.

Think of an alarm system, where the state logic can be opinionated, but the actions (e.g. notifications) should be freely definable.

Example:

```js
{
  states: {
    idle: {},
    arming: {
      states: {
        waiting: {},
        ready: {},
      },
    },
    armed: {},
    alarm: {},
}
```

I would like the user to be able to call e.g. `.withConfig({ actions: { 'arming.waiting.entry': () => notify('You have 1 minute left before the alarm system is armed!') }})`.

I could certainly do it by having every possible action defined explicitly in the machine config, as in:

```js
{
  states: {
    idle: {
      entry: ['idle.entry'],
      exit: ['idle.exit'],
    },
    arming: {
      states: {
        waiting: {
          entry: ['arming.waiting.entry'],
          exit: ['arming.waiting.exit'],
        },
        ready: {
          entry: ['arming.ready.entry'],
          exit: ['arming.ready.exit'],
        },
      },
      entry: ['arming.entry'],
      exit: ['arming.exit'],
    },
    armed: {
      entry: ['armed.entry'],
      exit: ['armed.exit'],
    },
    alarm: {
      entry: ['alarm.entry'],
      exit: ['alarm.exit'],
    },
}
```
 
But as you can see that quickly becomes *very* boilerplatey and error prone. And the machine is more complicated than in this simple example.

What would be the recommended way to do this in a more programmatically automated way? I have been thinking about using `.onTransition()`, but it seems this could only cover the `entry` actions properly? I could also expose a function that takes a machine config (a pojo like in the first code snipped above), and returns the augmented one (with all actions boilerplate added). But I wonder if there is a more straightforward way, that I have missed so far?

**Comment by davidkpiano** [comment, marked as answer]
For entry actions, yes, you should use `.onTransition()` or `.subscribe()`.

However, when reacting to changes (executing actions) in the state machine, the consumer should be unaware of the implementation details of the machine.

What if you decide to nest some of these states, add parallel states, or rename the keys? The actions should still be executed, but any entry/exit action that is tied to the state node path will break ⚠️.

That's why, although it seems verbose, IMO it's best to provide named entry/exit actions explicitly.

**Comment by simonihmig** [comment]
Yeah, that makes sense actually! It might be a bit hard to come up with good names for the actions, that are not just a duplicate of the state they are defined in. But we know already that naming things is hard.... Thanks for the quick feedback, appreciate it!
